Firestore/Example/Tests/Integration/API/FIRDatabaseTests.m

```diff
@@ -931,4 +932,175 @@ - (void)testWatchStreamReconnectsAfterIdle {
+  NSDictionary<NSString *, id> *initialData = @{@"key1" : @"value1"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // go offline for the rest of this test
+  [self disableNetwork];
+
+  // update the doc (though don't wait for a server response. We're offline; so
+  // that ain't happening!). This allows us to further distinguished cached vs
+  // server responses below.
+  NSDictionary<NSString *, id> *newData = @{@"key2" : @"value2"};
+  [doc setData:newData completion:^(NSError *_Nullable error) {
+    XCTAssertTrue(false, "Because we're offline, this should never occur.");
+  }];
+
+  // get doc and ensure it exists, *is* from the cache, and matches the
+  // newData.
+  FIRDocumentSnapshot *result = [self readDocumentForRef:doc getOptions:[FIRGetOptions fromCache]];
+  XCTAssertTrue(result.exists);
+  XCTAssertTrue(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, newData);
+}
+
+- (void)testGetDocumentWhileOnlineServerOnly {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key" : @"value"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // get doc and ensure that it exists, is *not* from the cache, and matches
+  // the initialData.
+  FIRDocumentSnapshot *result = [self readDocumentForRef:doc getOptions:[FIRGetOptions fromServer]];
+  XCTAssertTrue(result.exists);
+  XCTAssertFalse(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, initialData);
+}
+
+- (void)testGetDocumentWhileOfflineServerOnly {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key1" : @"value1"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // go offline for the rest of this test
+  [self disableNetwork];
+
+  // update the doc (though don't wait for a server response. We're offline; so
+  // that ain't happening!). This allows us to further distinguished cached vs
+  // server responses below.
+  NSDictionary<NSString *, id> *newData = @{@"key2" : @"value2"};
+  [doc setData:newData completion:^(NSError *_Nullable error) {
+    XCTAssertTrue(false, "Because we're offline, this should never occur.");
+  }];
+
+  // attempt to get doc and ensure cannot be retreived
+  // newData.
+  XCTestExpectation *failedGetDocCompletion = [self expectationWithDescription:@"failedGetDoc"];
+  // TODO: force server only
+  [doc getDocumentWithCompletion:^(FIRDocumentSnapshot *snapshot, NSError *error) {
+    XCTAssertNotNil(error);
+    // TODO: Ensure error is reasonable. Something like this but with different values:
+    //  XCTAssertEqualObjects(error.domain, FIRFirestoreErrorDomain);
+    //  XCTAssertEqual(error.code, FIRFirestoreErrorCodeNotFound);
+    [failedGetDocCompletion fulfill];
+  } getOptions:[FIRGetOptions fromServer]];
+  [self awaitExpectations];
+}
+
+- (void)testGetNonExistingDocWhileOnlineWithDefaultGetOptions {
```

Comment:
This is `testCanRetrieveDocumentThatDoesNotExist`.

Reply:
See above. (Also applies to method immediately below.)